File: src/main/java/dark/leech/text/enities/PluginEntity.java
package dark.leech.text.enities;

import com.google.gson.annotations.SerializedName;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * Entity representing a plugin with metadata and getter classes. Used for plugin management,
 * serialization, and updates.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class PluginEntity {

    @SerializedName("uuid")
    private String uuid;

    @SerializedName("name")
    private String name;

    @SerializedName("version")
    private Double version;

    @SerializedName("url")
    private String url;

    @SerializedName("language")
    private String language;

    @SerializedName("icon")
    private String icon;

    @SerializedName("source")
    private String source;

    @SerializedName("regex")
    private String regex;

    @SerializedName("author")
    private String author;

    @SerializedName("describe")
    private String describe;

    @SerializedName("group")
    private String group;

    @SerializedName("data")
    private String data;

    @Builder.Default private boolean supportUpdate = false;

    @SerializedName("chap")
    private String chapGetter;

    @SerializedName("toc")
    private String tocGetter;

    @SerializedName("page")
    private String pageGetter;

    @SerializedName("gen")
    private String genGetter;

    @SerializedName("search")
    private String searchGetter;

    @SerializedName("detail")
    private String detailGetter;

    @Builder.Default private boolean checked = false;

    /**
     * Apply all fields from another entity, marking this entity as checked. Used for updating
     * plugin metadata from remote sources.
     */
    public void apply(PluginEntity entity) {
        this.uuid = entity.uuid == null ? java.util.UUID.randomUUID().toString() : entity.uuid;
        this.name = entity.name;
        this.version = entity.version;
        this.url = entity.url;
        this.language = entity.language;
        this.icon = entity.icon;
        this.source = entity.source;
        this.regex = entity.regex;
        this.author = entity.author;
        this.describe = entity.describe;
        this.group = entity.group;
        this.data = entity.data;
        this.supportUpdate = entity.supportUpdate;
        this.chapGetter = entity.chapGetter;
        this.tocGetter = entity.tocGetter;
        this.pageGetter = entity.pageGetter;
        this.genGetter = entity.genGetter;
        this.searchGetter = entity.searchGetter;
        this.detailGetter = entity.detailGetter;
        this.checked = true;
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/entities/PluginEntity.kt
package dev.haipham22.leechtext.entities

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

/** Plugin manifest + getter scripts (port từ enities/PluginEntity.java). */
@Serializable
data class PluginEntity(
    @SerialName("uuid") var uuid: String? = null,
    @SerialName("name") var name: String? = null,
    @SerialName("version") var version: Double? = null,
    @SerialName("url") var url: String? = null,
    @SerialName("language") var language: String? = null,
    @SerialName("icon") var icon: String? = null,
    @SerialName("source") var source: String? = null,
    @SerialName("regex") var regex: String? = null,
    @SerialName("author") var author: String? = null,
    /** Tag vd "nsfw" — badge cảnh báo trong UI nguồn. */
    @SerialName("tag") var tag: String? = null,
    /** Ưu tiên khi nhiều plugin khớp 1 URL — cao thắng (vBook priority). */
    @SerialName("priority") var priority: Int = 0,
    @SerialName("describe") var describe: String? = null,
    @SerialName("group") var group: String? = null,
    @SerialName("data") var data: String? = null,
    var supportUpdate: Boolean = false,
    @SerialName("chap") var chapGetter: String? = null,
    @SerialName("toc") var tocGetter: String? = null,
    @SerialName("page") var pageGetter: String? = null,
    @SerialName("gen") var genGetter: String? = null,
    @SerialName("search") var searchGetter: String? = null,
    @SerialName("home") var homeGetter: String? = null,
    @SerialName("genre") var genreGetter: String? = null,
    @SerialName("tab") var tabGetter: String? = null,
    @SerialName("detail") var detailGetter: String? = null,
    // Toàn bộ src/*.js của plugin zip (filename -> content) — cho sandbox load() in-memory
    @SerialName("extra_scripts") var extraScripts: Map<String, String>? = null,
    @SerialName("config") var config: Map<String, String>? = null,
    /** Spec config cho UI: key → (title, subtitle, default, values, mode, format). */
    @SerialName("config_spec") var configSpec: Map<String, PluginConfigSpec>? = null,
    var checked: Boolean = false,
) {
    /** Copy toàn bộ field từ entity khác, đánh dấu checked (update metadata từ remote). */
    @OptIn(ExperimentalUuidApi::class)
    fun apply(entity: PluginEntity) {
        uuid = entity.uuid ?: Uuid.random().toString()
        name = entity.name
        version = entity.version
        url = entity.url
        language = entity.language
        icon = entity.icon
        source = entity.source
        regex = entity.regex
        author = entity.author
        describe = entity.describe
        group = entity.group
        data = entity.data
        supportUpdate = entity.supportUpdate
        chapGetter = entity.chapGetter
        tocGetter = entity.tocGetter
        pageGetter = entity.pageGetter
        genGetter = entity.genGetter
        searchGetter = entity.searchGetter
        detailGetter = entity.detailGetter
        checked = true
    }
}

/** Spec 1 entry config plugin (vBook extension-api) — dùng render UI cấu hình. */
@Serializable
data class PluginConfigSpec(
    var title: String? = null,
    var subtitle: String? = null,
    var default: String? = null,
    var values: List<String> = emptyList(),
    var mode: String? = null,
    var format: String? = null,
)
